release-19.1: storage: kick slow closed timestamps being used by rangefeed

danhhz opened the pull request:
Backport 1/1 commits from #36684 and 1/1 commits from #36772.

/cc @cockroachdb/release

---

Changefeeds and, more proximately, RangeFeeds need closed timestamps to
keep advancing to make progress, but sometimes we have bugs that lead to
closed timestamps getting stuck. We should and will fix the bugs, but
given that even one stuck closed timestamp will stall changefeed-wide
progress, this commit introduces a fallback to try kicking the closed
timestamp when it's behind. Since these are unexected, we also log them
and keep a metric.

Touches #35142

Release note (bug fix): Made `CHANGEFEED`s more resilient to a class of
bugs which manifest as stalls.

---

storage: limit the amount of concurrent work started by the rangefeed nudger

In #36684, we introduced a closed timestamp "nudge" whenever a RangeFeed
finds that the closed timestamp has fallen more than some threshold
behind realtime. This commit limits the number of those that can be run
concurrently, to protect against the case of a system-wide slowdown on
closed timestamps, which would otherwise potentially launch a launching
a potentially huge number of lease acquisitions all at once.

Release note: None

danhhz committed:
storage: nudge slow closed timestamps being used by rangefeed

Changefeeds and, more proximately, RangeFeeds need closed timestamps to
keep advancing to make progress, but sometimes we have bugs that lead to
closed timestamps getting stuck. We should and will fix the bugs, but
given that even one stuck closed timestamp will stall changefeed-wide
progress, this commit introduces a fallback to try nudging the closed
timestamp when it's behind. Since these are unexected, we also log them
and keep a metric.

Touches #35142

Release note (bug fix): Made `CHANGEFEED`s more resilient to a class of
bugs which manifest as stalls.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36768)
<!-- Reviewable:end -->

nvanbenschoten commented:
I left a few more comments on #36684. The only thing that concerns me enough to be hesitant about merging this is that we're not limiting the number of goroutines calling `ensureClosedTimestampStarted`. This seems bad. It means that the second a cluster loses connectivity, each node will launch `O(num replicas with rangefeeds)` goroutines.

danhhz committed:
storage: limit the amount of concurrent work started by the rangefeed nudger

In #36684, we introduced a closed timestamp "nudge" whenever a RangeFeed
finds that the closed timestamp has fallen more than some threshold
behind realtime. This commit limits the number of those that can be run
concurrently, to protect against the case of a system-wide slowdown on
closed timestamps, which would otherwise potentially launch a launching
a potentially huge number of lease acquisitions all at once.

Release note: None

danhhz commented:
@nvanbenschoten this is ready to go, I added the folowup PR, which is now borsing

danhhz commented:
Flake was a TestVet OOM. I queued up another

nvb reviewed (approved):
LGTM

danhhz commented:
TFTR!
